matrix/tasks/deploy.py

```diff
@@ -1,7 +1,83 @@
+from matrix.utils import execute_process
+from matrix.model import TestFailure
 
+import yaml
 
-async def deploy(context, rule, task, event=None):
-    rule.log.info("Deploying %s", context.config.path)
+
+async def libjuju(context, rule):
+    """
+    Deploy a bundle or charm via libjuju.
+
+    """
     await context.juju_model.deploy(str(context.config.path))
+
+
+async def get_controller_name(controller_id, log):
+    if controller_id.startswith('controller-'):
+        controller_id = controller_id[11:]
+
+    success, controllers, _ = await execute_process(
+        ["juju", "list-controllers", "--format", "yaml"],
```

Comment:
@johnsca Yes. That's why I was asking about the best way to do it in python-libjuju. You told me there wasn't a way, which I suppose means that I should read the code, rather than trusting you :-p